### PR TITLE
TST / DEBUG: better warnings and tests when facing singular hessian problems

### PR DESCRIPTION
**Pull request by ogrisel**
Here is an attempt (failed or successful depending on the viewpoint) at improving the tests for the fallback inner solver for the singular Hessian case in the context of the `"newton-cholesky"` solver introduced in https://github.com/scikit-learn/scikit-learn/pull/23314.

I think in it's current state, it's pretty useless:

- either with `lstsq` or `indefinite_factorization`, the model never converges, all line searches fail, the deviance stays very high and we issue a ton of useless warnings.
- `lstsq` seems to be a bit faster but it's not important since neither method is helpful at making the model converge.

Possible things to try:

- a) do not try to do a linear search on a singular hessian solution and instead continue warning but instead do a simple gradient step with small learning rate. But then we are not sure how small a learning rate. Or maybe we could try to run the line search with the raw gradient direction?
- b) when facing a singular hessian, just stop the solver with a helpful convergence warning that suggests less collinear features or stronger regularization
- c) alternatively, we could warn and try to fit the linear model again with strong regularization automatically (without ever trying to find a solution for the first encountered singular hessian problem)

Note that the suggestion to slightly increase the regularization works as shown in the last section of the test.

### CODE REVIEW

File: sklearn/linear_model/_glm/glm.py [hunk 1]
```diff
@@ -316,8 +326,9 @@ def line_search(self, X, y, sample_weight):
             t *= beta
         else:
             warnings.warn(
-                f"Line search of Newton solver {self.__class__.__name__} did not "
-                "converge after 21 line search refinement iterations.",
+                f"Line search of Newton solver {self.__class__.__name__} at iteration"
+                f" #{self.iteration} did not converge after 21 line search refinement"
```
Comment (ogrisel) [inline sklearn/linear_model/_glm/glm.py:330]: I think it can be helpful to include the iteration number in the warning but I am not sure if we should use the one-based convention `"#{self.iteration}"` or the 0-based convention used elsewhere in the code with `#{self.iteration - 1}`.

File: sklearn/linear_model/_glm/glm.py [hunk 2]
```diff
@@ -31,37 +31,47 @@
 from .._linear_loss import LinearModelLoss
 
 
-def _solve_singular(H, g):
+def _solve_singular(H, g, method="lstsq"):
```
Comment (ogrisel) [inline sklearn/linear_model/_glm/glm.py:34]: I decided to privately expose the `method` param to quickly switch between the 2 approaches to make it easier to investigate the singular Hessian problem but as I found that neither seems to be helpful I do not know which to keep. If we can actually make use of the `indefinite_factorization` variant, I think we should unittest it individually.

File: sklearn/linear_model/_glm/tests/test_glm.py [hunk 3]
```diff
@@ -794,14 +794,40 @@ def test_linalg_warning_with_newton_solver(global_random_seed):
...
-        "singular hessian matrix. "
+        "singular hessian matrix"
+        ".*"
+        "increasing the regularization strength may resolve this issue"
     )
-    with pytest.warns(scipy.linalg.LinAlgWarning, match=msg):
-        PoissonRegressor(solver="newton-cholesky", alpha=0.0).fit(X_colinear, y)
+    with pytest.warns(scipy.linalg.LinAlgWarning, match=msg) as rec:
+        reg = PoissonRegressor(solver="newton-cholesky", alpha=0.0).fit(X_colinear, y)
+    # XXX: this cases raises a ton of convergence warnings:
```
Comment (ogrisel) [inline sklearn/linear_model/_glm/tests/test_glm.py]:
```suggestion
    # XXX: this case raises a ton of convergence warnings:
```

File: sklearn/linear_model/_glm/tests/test_glm.py [hunk 4]
```diff
@@ -794,14 +794,40 @@ def test_linalg_warning_with_newton_solver(global_random_seed):
     y = rng.normal(size=X_orig.shape[0])
     y[y < 0] = 0.0
 
+    # No warning raised on well-conditioned design, even without regularization.
     with warnings.catch_warnings():
         warnings.simplefilter("error")
-        # No warning raised on well-conditioned design
-        PoissonRegressor(solver="newton-cholesky", alpha=0.0).fit(X_orig, y)
+        reg = PoissonRegressor(solver="newton-cholesky", alpha=0.0).fit(X_orig, y)
+        reference_deviance = mean_poisson_deviance(y, reg.predict(X_orig))
```
Comment (ogrisel) [inline sklearn/linear_model/_glm/tests/test_glm.py]:
```suggestion
    reference_deviance = mean_poisson_deviance(y, reg.predict(X_orig))
```

**Comment by lorentzenchr** [comment]
Nocedal & Wright write in Chapter 3.4:
> As this discussion shows, there is a great deal of freedom in devising modification strategies, and there is currently no agreement on which strategy is best.

Let's see, if we find a satisfying solution.

**Comment by ogrisel** [comment]
I moved to a new place and my copy of N & W is still somewhere buried in a cardboard box :)

**Comment by lorentzenchr** [comment]
I finally went with option **a) simple gradient steps**, see https://github.com/scikit-learn/scikit-learn/pull/23314#issuecomment-1152895761.
I think I can close with https://github.com/scikit-learn/scikit-learn/pull/23314/commits/8a108bb88969341514d0730dd6b68dace5c269b3 and https://github.com/scikit-learn/scikit-learn/pull/23314/commits/82287af585a89d68b1dcf9190b37663c40b7327c.
